src/SFML/Graphics/Image.cpp

```diff
@@ -28,85 +28,73 @@
 #include <SFML/Graphics/Image.hpp>
 #include <SFML/Graphics/ImageLoader.hpp>
 #include <SFML/System/Err.hpp>
+
+
 #ifdef SFML_SYSTEM_ANDROID
 #include <SFML/System/Android/ResourceStream.hpp>
 #endif
 #include <algorithm>
 #include <cstring>
+#include <optional>
 #include <ostream>
+#include <utility>
 
 
 namespace sf
 {
 ////////////////////////////////////////////////////////////
-void Image::create(const Vector2u& size, const Color& color)
+Image::Image(Vector2u size, std::vector<std::uint8_t>&& pixels) : m_size(size), m_pixels(std::move(pixels))
 {
-    if (size.x && size.y)
-    {
-        // Create a new pixel buffer first for exception safety's sake
-        std::vector<std::uint8_t> newPixels(static_cast<std::size_t>(size.x) * static_cast<std::size_t>(size.y) * 4);
+}
 
-        // Fill it with the specified color
-        std::uint8_t* ptr = newPixels.data();
-        std::uint8_t* end = ptr + newPixels.size();
-        while (ptr < end)
-        {
-            *ptr++ = color.r;
-            *ptr++ = color.g;
-            *ptr++ = color.b;
-            *ptr++ = color.a;
-        }
+////////////////////////////////////////////////////////////
+std::optional<Image> Image::create(const Vector2u& size, const Color& color)
+{
+    if (size.x == 0 || size.y == 0)
+    {
+        return std::nullopt;
+    }
```

Comment:
Maybe this is something you already discussed elsewhere, apologies if so. But I don't think this is an appropriate use of `std::optional`. If you have an invariant that `Image` must always have non-zero width and height, then calling `Image::create({0, 0})` is a pre-condition violation, i.e., a bug. It should assert/terminate/throw, not return `std::nullopt`. It is always the caller's responsibility to not write bugs.

`std::nullopt` should be a valid return value if you asked something sensible and within the spec, but the function was unable to comply. This could be because the file wasn't found, or wasn't readable, or was corrupted, or because there wasn't enough memory, etc., so it's absolutely fine for all the `loadFrom*` functions. 

But "invalid input" by definition isn't within the spec, so shouldn't be a cause for returning `std::nullopt`. The same rationale applies to the first check in `Image::create(const Vector2u& size, const std::uint8_t* pixels)`.

A more appropriate use of `std::optional` here would be to catch the allocation failure when creating the `std::vector` pixel buffer, and return `std::nullopt` when out-of-memory. In my opinion, allocation failure is often under-rated, and can absolutely happen (especially here since you don't know your user's hardware resources and image size requirements). But that's probably part of a larger debate, and depends how serious you want to be about handling all failure modes.